powertools-batch/src/main/java/software/amazon/lambda/powertools/batch/handler/DynamoDbBatchMessageHandler.java

```diff
@@ -46,35 +49,60 @@ public DynamoDbBatchMessageHandler(Consumer<DynamodbEvent.DynamodbStreamRecord>
 
     @Override
     public StreamsEventResponse processBatch(DynamodbEvent event, Context context) {
-        List<StreamsEventResponse.BatchItemFailure> batchFailures = new ArrayList<>();
+        StreamsEventResponse response = StreamsEventResponse.builder().withBatchItemFailures(new ArrayList<>()).build();
 
-        for (DynamodbEvent.DynamodbStreamRecord record : event.getRecords()) {
-            try {
+        for (DynamodbEvent.DynamodbStreamRecord streamRecord : event.getRecords()) {
+            processBatchItem(streamRecord, context).ifPresent(batchItemFailure -> response.getBatchItemFailures().add(batchItemFailure));
+        }
 
-                rawMessageHandler.accept(record, context);
-                // Report success if we have a handler
-                if (this.successHandler != null) {
-                    this.successHandler.accept(record);
-                }
-            } catch (Throwable t) {
-                String sequenceNumber = record.getDynamodb().getSequenceNumber();
-                LOGGER.error("Error while processing record with id {}: {}, adding it to batch item failures",
-                        sequenceNumber, t.getMessage());
-                LOGGER.error("Error was", t);
-                batchFailures.add(new StreamsEventResponse.BatchItemFailure(sequenceNumber));
-
-                // Report failure if we have a handler
-                if (this.failureHandler != null) {
-                    // A failing failure handler is no reason to fail the batch
-                    try {
-                        this.failureHandler.accept(record, t);
-                    } catch (Throwable t2) {
-                        LOGGER.warn("failureHandler threw handling failure", t2);
-                    }
+        return response;
+    }
+
+
+    @Override
+    public StreamsEventResponse processBatchInParallel(DynamodbEvent event, Context context) {
+        MultiThreadMDC multiThreadMDC = new MultiThreadMDC();
+
+        List<StreamsEventResponse.BatchItemFailure> batchItemFailures = event.getRecords()
+                .parallelStream() // Parallel processing
+                .map(eventRecord -> {
+                    multiThreadMDC.copyMDCToThread(Thread.currentThread().getName());
```

Comment:
What is using ThreadLocal in tracing ?

Serialization yes, with this ObjectMapper that we need to think about seriously. Tracing, I'm not sure...

Reply:
Tracing will use it for span propagation I reckon

Reply:
X-Ray documents how we can explicitly [pass segment context between threads](https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-java.html#xray-sdk-java-multithreading). The trick'll be that we need to do this without binding the two libraries together.

Reply:
Good catch... but we probably don't want to have the X-Ray dependency (or even the PT tracing one) in batch.
Then how can we do ? Some dirty reflection:

```java
    Class<?> awsXRayClass =
                Class.forName("com.amazonaws.xray.AWSXRay", true, this.getClass().getClassLoader());
    Method getGlobalRecorder = awsXRayClass.getMethod("getGlobalRecorder");
    Object awsXRayRecorder = getGlobalRecorder.invoke(null);
    Method getTraceEntity = awsXRayRecorder.getClass().getMethod("getTraceEntity");
    Object segment = getTraceEntity.invoke(awsXRayRecorder);
    ...
```

Reply:
There is no such service thing for XRay, and even if we created one, we still need the XRay classes/interfaces...

It makes me think of this issue #1505 that @mriccia created. Should we introduce XRay in our code, even if we don't know if people use tracing? `aws-java-sdk-xray-1.12.228.jar` is 611 kb...

Reply:
Based in my tests, I don't see any problem with Tracing:

![image](https://github.com/aws-powertools/powertools-lambda-java/assets/117538/ebebf1f1-4865-467b-a872-17cc01dfbe37)

Works both with the `@Tracing` annotation on processMessage method and with the `TracingUtils.withSubsegment("processMessage", (subsegment -> {...}));`

Could also add annotations and metadata.

Reply:
>  Should we introduce XRay in our code, even if we don't know if people use tracing?

I think no, because we are actively decoupling tracing from x-ray at the moment (#1468, @jasoniharris and I are working on this now). If anything we'd couple around some lightweight interface in `powertools-tracing`. Maybe that's our soln? 

> Works both with the @Tracing annotation on processMessage method and with the 

I'm not sure this shows it's working. Can you show me that the segments "outside of" the batch handler - which are definitely running in another thread - are tied together with these?

Reply:
To wrap this thread up, we have seen that we get spans disconnecting from each other when throwing work into the ThreadPool. We've raised #1671 to deal with this asynchronously